File: packages/core/src/parse.ts
import color from 'better-color-tools';

export interface ColorToken extends TokenBase<string> {
  type: 'color';
}

export interface ConicGradientToken extends TokenBase<string> {
  type: 'conic-gradient';
}

export interface CubicBezierToken extends TokenBase<[number, number, number, number]> {
  type: 'cubic-bezier';
}

export interface DimensionToken extends TokenBase<string> {
  type: 'dimension';
}

export interface FontToken extends TokenBase<string[]> {
  type: 'font';
}

export interface FileToken extends TokenBase<string> {
  type: 'file';
}

export type Group = {
  metadata?: Record<string, unknown>;
} & {
  [childNode: string]: TokenOrGroup;
};

export interface LinearGradientToken extends TokenBase<string> {
  type: 'linear-gradient';
}

export type Mode<T = string> = Record<string, T>;

export interface RadialGradientToken extends TokenBase<string> {
  type: 'radial-gradient';
}

export interface ShadowToken extends TokenBase<string[]> {
  type: 'shadow';
}

export interface TokenBase<T = string> {
  /** User-friendly name */
  name?: string;
  /** Token description */
  description?: string;
  /** Token value */
  value: T;
  /** Mode variants */
  mode: Mode<T>;
}

export type Token =
  | ColorToken
  | ConicGradientToken
  | CubicBezierToken
  | DimensionToken
  | FileToken
  | FontToken
  | LinearGradientToken
  | RadialGradientToken
  | ShadowToken
  | URLToken;

export type TokenType = Token['type'];

export type TokenOrGroup = Group | Token;

export interface URLToken extends TokenBase<string> {
  type: 'url';
  value: string;
}

export interface ParsedMetadata {
  name?: string;
  version?: string;
  metadata?: Record<string, unknown>;
}

export interface ParseResult {
  errors?: string[];
  warnings?: string[];
  result: {
    metadata: ParsedMetadata;
    tokens: ParsedToken[];
  };
}

export type ParsedToken = { id: string } & Token;

export interface Schema {
  name?: string;
  version?: string;
  metadata?: Record<string, unknown>;
  tokens: Record<string, TokenOrGroup>;
}

const VALID_TOP_LEVEL_KEYS = new Set(['name', 'version', 'metadata', 'tokens']);
const ALIAS_RE = /^\{.*\}$/;

export function parse(schema: Schema): ParseResult {
  const errors: string[] = [];
  const warnings: string[] = [];
  const result: ParseResult = { result: { metadata: {}, tokens: [] } };
  if (!schema || typeof schema !== 'object' || Array.isArray(schema)) {
    errors.push(`Invalid schema type. Expected object, received "${Array.isArray(schema) ? 'Array' : typeof schema}"`);
    result.errors = errors;
    return result;
  }

  // 1. check top-level
  for (const k of Object.keys(schema)) {
    if (VALID_TOP_LEVEL_KEYS.has(k)) {
      if (k !== 'tokens') (result.result.metadata as any)[k] = (schema as any)[k];
    } else {
      errors.push(`Invalid top-level name "${k}". Place arbitrary data inside "metadata".`);
    }
  }
  if (errors.length) {
    result.errors = errors;
    return result;
  }
  if (!schema.tokens || typeof schema.tokens !== 'object' || !Object.keys(schema.tokens).length) {
    errors.push('"tokens" is empty!');
    result.errors = errors;
    return result;
  }

  interface InheritedGroup {
    type?: TokenType;
    modes: string[];
  }

  // 2. collect tokens
  const tokens: Record<string, ParsedToken> = {};
  function walk(node: TokenOrGroup, chain: string[] = [], group: InheritedGroup = { modes: [] }): void {
    for (const [k, v] of Object.entries(node)) {
      if (!v || Array.isArray(v) || typeof v !== 'object') {
        errors.push(`${k}: unexpected token format "${v}"`);
        continue;
      }
      if (k.includes('.') || k.includes('#')) {
        errors.push(`${k}: invalid name. Names can’t include "." or "#".`);
        continue;
      }

      // token
      const isToken = v.type || (group.type && !!v.value);
      if (isToken) {
        const id = chain.concat(k).join('.');
        const nodeType: TokenType = v.type || group.type;
        if (!v.value) {
          errors.push(`${id}: missing value`);
          continue;
        }
        const token = v as Token;
        let mode = token.mode || {};
        if (typeof mode !== 'object' || Array.isArray(mode)) {
          errors.push(`${id}: "mode" must be an object`);
          mode = {};
        }
        if (group.modes.length) {
          for (const modeID of group.modes) {
            if (!mode[modeID]) errors.push(`${id}: missing mode "${modeID}" set on parent group`);
          }
        }

        for (const [modeID, modeValue] of Object.entries(mode)) {
          if (!checkStrVal(modeValue)) errors.push(`${id}#${modeID}: bad value "${modeValue}"`);
        }
        tokens[id] = {
          id,
          ...(token as any),
          type: nodeType,
          mode: mode as any,
        };
      }
      // group
      else {
        const { metadata, ...groupTokens } = v; // "metadata" only reserved word on group
        const nextGroup = { ...group };
        if (metadata) {
          if (metadata.type) nextGroup.type = metadata.type;
          if (Array.isArray(metadata.modes)) nextGroup.modes = metadata.modes;
        }
        if (Object.keys(groupTokens).length) {
          walk(groupTokens, [...chain, k], nextGroup);
        }
      }
    }
  }
  walk(schema.tokens);
  if (errors.length) {
    result.errors = errors;
    return result;
  }

  // 3. resolve aliases
  const values: Record<string, unknown> = {};
  // 3a. pass 1: gather all IDs & values
  for (const token of Object.values(tokens)) {
    values[token.id] = token.value;
    for (const [k, v] of Object.entries(token.mode)) {
      values[`${token.id}#${k}`] = v;
    }
  }
  // 3b. pass 2: resolve simple aliases
  aliasLoop: while (Object.values(values).some((t) => typeof t === 'string' && ALIAS_RE.test(t))) {
    for (const [k, v] of Object.entries(values)) {
      if (typeof v !== 'string' || !ALIAS_RE.test(v)) continue;
      const id = v.substring(1, v.length - 1);
      if (!values[id]) {
        errors.push(`${k}: can’t find ${v}`);
        break aliasLoop;
      }
      // check for circular references
      const ref = values[id] as string;
      if (typeof ref === 'string' && ALIAS_RE.test(ref) && id === ref.substring(1, ref.length - 1)) {
        errors.push(`${k}: can’t reference circular alias ${v}`);
        break aliasLoop;
      }
      values[k] = values[id];
    }
  }
  if (errors.length) {
    result.errors = errors;
    return result;
  }
  // 3c. pass 3: resolve embedded aliases from simple aliases
  while (
    Object.values(values).some(
      (t) => (typeof t === 'string' && findAliases(t).length) || (Array.isArray(t) && t.some((v) => typeof v === 'string' && findAliases(v).length))
    )
  ) {
    for (const [k, v] of Object.entries(values)) {
      if (typeof v === 'string') {
        let value = v;
        for (const alias of findAliases(v)) {
          const aliasedID = alias.substring(1, alias.length - 1);
          values[k] = value.replace(alias, values[aliasedID] as any);
        }
      }
      if (Array.isArray(v) && v.every((s) => typeof s === 'string')) {
        values[k] = v.map((s) => {
          let value = s;
          for (const alias of findAliases(s)) {
            const aliasedID = alias.substring(1, alias.length - 1);
            value = value.replace(alias, values[aliasedID]);
          }
          return value;
        });
      }
    }
  }

  // 4. validate values & replace aliases
  for (const id of Object.keys(tokens)) {
    const token = tokens[id];
    try {
      switch (token.type) {
        // string tokens can all be validated together
        case 'dimension':
        case 'file':
        case 'linear-gradient':
        case 'radial-gradient':
        case 'conic-gradient': {
          const val = values[id] as string;
          // ✔ valid string
          if (checkStrVal(val)) tokens[id].value = val;
          // ✘ invalid string
          else errors.push(`${id}: bad value "${val}"`);
          for (const modeID of Object.keys(token.mode)) {
            const modeVal = values[`${id}#${modeID}`];
            // ✔ valid mode
            if (checkStrVal(modeVal)) tokens[id].mode[modeID] = modeVal as string;
            // ✘ invalid mode
            else errors.push(`${id}: bad value "${modeVal}"`);
          }
          break;
        }
        case 'color': {
          const val = values[id] as ColorToken['value'];
          // ✔ valid color
          if (checkColor(val)) tokens[id].value = val;
          // ✘ invalid color
          else errors.push(`${id}: invalid color "${val}"`);
          for (const modeID of Object.keys(token.mode)) {
            const modeVal = values[`${id}#${modeID}`] as ColorToken['value'];
            // ✔ valid mode
            if (checkColor(modeVal)) tokens[id].mode[modeID] = modeVal;
            // ✘ invalid mode
            else errors.push(`${id}: invalid color "${modeVal}"`);
          }
          break;
        }
        case 'font': {
          const rawVal = values[id];
          const val = Array.isArray(rawVal) ? rawVal : [rawVal];
          // ✔ valid font
          if (checkFont(val)) tokens[id].value = val;
          // ✘ invalid font
          else errors.push(`${id}: expected string or array of strings, received ${typeof token.value}`);
          for (const modeID of Object.keys(token.mode)) {
            const rawModeVal = values[`${id}#${modeID}`];
            const modeVal = Array.isArray(rawModeVal) ? rawModeVal : [rawModeVal];
            // ✔ valid mode
            if (checkFont(modeVal)) tokens[id].mode[modeID] = modeVal;
            // ✘ invalid mode
            else errors.push(`${id}: expected string or array of strings, received ${typeof rawModeVal}`);
          }
          break;
        }
        case 'cubic-bezier': {
          const val = values[id] as CubicBezierToken['value'];
          // ✔ valid cubic-bezier
          if (checkCubicBezier(val)) {
            val[0] = Math.max(0, Math.min(1, val[0]));
            val[2] = Math.max(0, Math.min(1, val[2]));
            tokens[id].value = val;
          }
          // ✘ invalid cubic-bezier
          else errors.push(`${id}: expected [x1, y1, x2, y2], received "${val}"`);
          for (const modeID of Object.keys(token.mode)) {
            const modeVal = values[`${id}#${modeID}`] as CubicBezierToken['value'];
            // ✔ valid mode
            if (checkCubicBezier(modeVal)) {
              modeVal[0] = Math.max(0, Math.min(1, modeVal[0]));
              modeVal[2] = Math.max(0, Math.min(1, modeVal[2]));
              tokens[id].mode[modeID] = modeVal;
            }
            // ✘ invalid mode
            else errors.push(`${id}: expected [x1, y1, x2, y2], received "${modeVal}"`);
          }
          break;
        }
        case 'url': {
          const val = values[id] as URLToken['value'];
          // ✔ valid url
          if (checkURL(val)) tokens[id].value = val;
          // ✘ invalid url
          else errors.push(`${id}: invalid url "${val}" (if this is relative, use type: "file")`);
          for (const modeID of Object.keys(token.mode)) {
            const modeVal = values[`${id}#${modeID}`] as URLToken['value'];
            // ✔ valid mode
            if (checkURL(modeVal)) tokens[id].mode[modeID] = modeVal;
            // ✘ invalid mode
            else errors.push(`${id}: invalid url "${modeVal}" (if this is relative, use type: "file")`);
          }
          break;
        }
        case 'shadow': {
          const val = values[id] as ShadowToken['value'];
          // ✔ valid shadow (validate before aliasing, as aliases may be inside array)
          if (checkShadow(val)) tokens[id].value = val;
          // ✘ invalid shadow
          else errors.push(`${id}: expected array, received "${token.value}"`);
          for (const modeID of Object.keys(token.mode)) {
            const modeVal = values[`${id}#${modeID}`] as ShadowToken['value'];
            // ✔ valid mode
            if (checkShadow(modeVal)) tokens[id].mode[modeID] = modeVal;
            // ✘ invalid mode
            else errors.push(`${id}: expected array, received "${modeVal}"`);
          }
          break;
        }
        // custom/other
        default: {
          tokens[id].value = values[id] as any;
          for (const modeID of Object.keys((token as any).mode)) {
            (tokens[id] as any).mode[modeID] = values[`${id}#${modeID}`];
          }
          break;
        }
      }
    } catch (err: any) {
      errors.push(`${id}: ${err.message || err}`);
    }
  }

  // 4. return
  if (errors.length) result.errors = errors;
  if (warnings.length) result.warnings = warnings;
  result.result.tokens = Object.values(tokens);
  return result;
}

function checkColor(val: unknown): boolean {
  if (!val) return false;
  if (typeof val !== 'string') return false;
  try {
    color.from(val);
    return true;
  } catch {
    return false;
  }
}

function checkCubicBezier(val: unknown): boolean {
  if (!val) return false;
  return Array.isArray(val) && val.length === 4 && val.every((v) => typeof v === 'number' && !Number.isNaN(v));
}

function checkStrVal(val: unknown): boolean {
  return !!val && typeof val === 'string';
}

function checkFont(val: unknown): boolean {
  if (!val) return false;
  if (typeof val === 'string') return true;
  return Array.isArray(val) && val.every((v) => typeof v === 'string' && v.length);
}

function checkShadow(val: unknown): boolean {
  if (!val) return false;
  return Array.isArray(val) && val.every((v) => typeof v === 'string' && v.length);
}

function checkURL(val: unknown): boolean {
  if (!val || typeof val !== 'string') return false;
  try {
    new URL(val);
    return true;
  } catch {
    return false;
  }
}

/** given a string, find all {aliases} */
function findAliases(input: string): string[] {
  const matches: string[] = [];
  if (!input.includes('{')) return matches;
  let lastI = -1;
  for (let n = 0; n < input.length; n++) {
    switch (input[n]) {
      case '\\': {
        // if '\{' or '\}' encountered, skip
        if (input[n + 1] == '{' || input[n + 1] == '}') n += 1;
        break;
      }
      case '{': {
        lastI = n; // '{' encountered; keep going until '}' (below)
        break;
      }
      case '}': {
        if (lastI === -1) continue; // ignore '}' if no '{'
        matches.push(input.substring(lastI, n + 1));
        lastI = -1; // reset last index
        break;
      }
    }
  }
  return matches;
}

File: packages/core/src/parse.rs
extern crate json;
extern crate yaml_rust;
use lazy_static::lazy_static;
use regex::Regex;
use std::collections::{HashMap, HashSet};
use yaml_rust::{Yaml, YamlLoader};

// TODO: validate modes

lazy_static! {
    static ref HEX_RE: Regex = Regex::new(r"^#[0-9A-Fa-f]{3,8}$").unwrap();
    static ref URL_RE: Regex = Regex::new(r"^(http://|https://|//)").unwrap();
    static ref ALIAS_RE: Regex = Regex::new(r"^\$").unwrap();
}

// convert tokens.yaml into JSON string
pub fn parse_and_validate(raw_schema: &String) -> String {
    let mut schema = json::object! {
        result: json::object!{
            metadata: json::JsonValue::new_object(),
            tokens: json::JsonValue::new_array(),
        },
        errors: json::JsonValue::new_array()
    };

    let yaml_docs = YamlLoader::load_from_str(&raw_schema);

    // 0. if scan error, return error & exit
    match yaml_docs {
        Err(e) => {
            schema["errors"].push(format!("YAML parse error: {}", e));
            return schema.dump();
        }
        _ => {}
    }
    let yaml = &yaml_docs.unwrap()[0];

    // 1. validate top-level keys
    for (raw_k, raw_v) in yaml.as_hash().unwrap() {
        let k = raw_k.as_str().unwrap();
        match k {
            "name" | "version" | "metadata" => {
                schema["result"]["metadata"][k] = json_serialize(raw_v);
            }
            "tokens" => {
                if raw_v.as_hash().is_some() {
                    if raw_v.as_hash().unwrap().keys().len() == 0 {
                        schema["errors"].push("\"tokens\" is empty!");
                    }
                } else {
                    schema["errors"].push("invalid \"tokens\" format");
                }
            }
            _ => {
                let msg = format!(
                    "Invalid top-level name \"{}\". Place arbitrary data inside \"metadata\"",
                    k,
                );
                schema["errors"].push(msg);
            }
        }
    }
    if is_empty(&yaml["tokens"]) {
        schema["errors"].push("\"tokens\" is empty!");
    }
    if schema["errors"].len() > 0 {
        return schema.dump();
    }

    // 2. iterate through tokens
    fn walk(node: &Yaml, id: &String, obj: &mut json::JsonValue) {
        // group
        if is_empty(&node["type"]) || node["type"].as_str().is_none() {
            if node.as_hash().is_some() {
                for (k, v) in node.as_hash().unwrap() {
                    let next_id = format!("{}.{}", id, k.as_str().unwrap());
                    walk(v, &next_id, obj);
                }
            } else {
                obj["errors"].push(json::from(format!("syntax error on group \"{}\"", id)));
            }
        }
        // token
        else {
            let token = match node["type"].as_str().unwrap() {
                "color" => parse_color(node, id),
                "dimension" => parse_dimension(node, id),
                "font" => parse_font(node, id),
                "cubic-bezier" => parse_cubic_bezier(node, id),
                "file" => parse_str(node, id),
                "url" => parse_url(node, id),
                "shadow" => parse_shadow(node, id),
                "linear-gradient" => parse_str(node, id),
                "radial-gradient" => parse_str(node, id),
                "conic-gradient" => parse_str(node, id),
                _ => parse_str(node, id),
            };
            match token {
                Err(msgs) => {
                    for e in msgs {
                        obj["errors"].push(json::from(e));
                    }
                }
                Ok(t) => {
                    obj["result"]["tokens"].push(t);
                }
            }
        }
    }

    for (k, v) in yaml["tokens"].as_hash().unwrap() {
        walk(v, &k.as_str().unwrap().to_owned(), &mut schema);
    }

    if schema["errors"].len() > 0 {
        return schema.dump();
    }

    // 3. TODO: validate modes

    // 4. validate aliases
    let mut aliases: HashMap<String, String> = HashMap::new();
    fn follow_aliases(
        original_id: &String,
        next_node: &json::JsonValue,
        resolved: &mut HashMap<String, String>,
        scanned: &mut HashSet<String>,
        schema: &json::JsonValue,
    ) -> Result<String, Vec<String>> {
        // if already resolved, exit
        let id = next_node["id"].as_str().unwrap().to_owned();
        if resolved.contains_key(&id) {
            return Ok(resolved.get(&id).unwrap().to_owned());
        }

        // if this isn’t an alias, exit
        if next_node["value"].as_str().is_none()
            || !ALIAS_RE.is_match(next_node["value"].as_str().unwrap())
        {
            resolved.insert(original_id.to_owned(), id.to_owned());
            return Ok(id.to_owned());
        }

        let mut errors: Vec<String> = Vec::new();
        let alias = next_node["value"].as_str().unwrap().to_owned();
        let alias_id = &alias[1..alias.len()];

        // exit if missing alias
        let mut aliased: &json::JsonValue = &json::Null;
        for token in schema["result"]["tokens"].members() {
            if token["id"].as_str().unwrap() == alias_id {
                aliased = token;
                break;
            }
        }
        if aliased.is_null() {
            errors.push(format!("{}: {} not found", id, alias));
            return Err(errors.to_owned());
        }

        // exit if type mismatch
        if next_node["type"].as_str().unwrap() != aliased["type"].as_str().unwrap() {
            errors.push(format!(
                "{}: can’t use {} value for {}",
                id,
                aliased["type"].as_str().unwrap(),
                next_node["type"].as_str().unwrap()
            ));
            return Err(errors.to_owned());
        }

        // exit if circular reference
        if scanned.get(&id).is_some() {
            errors.push(format!(
                "{}: can’t reference circular alias {}",
                id.to_owned(),
                alias
            ));
            return Err(errors.to_owned());
        }

        // if next node is another alias, continue scanning
        if aliased["value"].as_str().is_some()
            && ALIAS_RE.is_match(&aliased["value"].as_str().unwrap())
        {
            scanned.insert(id); // prevent infinite loop
            return follow_aliases(original_id, aliased, resolved, scanned, schema);
        }

        // otherwise, finish & resolve
        resolved.insert(original_id.to_owned(), alias_id.to_owned());
        return Ok(alias_id.to_owned());
    }
    for token in schema["result"]["tokens"].members() {
        let mut scanned: HashSet<String> = HashSet::new();
        match follow_aliases(
            &token["id"].as_str().unwrap().to_owned(),
            token,
            &mut aliases,
            &mut scanned,
            &schema,
        ) {
            Ok(_) => {
                // aliases are valid! do nothing
            }
            Err(errors) => {
                for msg in errors {
                    schema["errors"].push(msg);
                }
                return schema.dump();
            }
        }
    }

    return schema.dump();
}

// convert Yaml to Json
pub fn json_serialize(node: &Yaml) -> json::JsonValue {
    if is_empty(node) {
        return json::from("null");
    }
    if node.is_array() {
        let mut items = json::JsonValue::new_array();
        for v in node.as_vec().unwrap() {
            items.push(json_serialize(v));
        }
        return items;
    }
    if node.as_hash().is_some() {
        let mut obj = json::JsonValue::new_object();
        for (k, v) in node.as_hash().unwrap() {
            obj[k.as_str().unwrap().to_owned()] = json_serialize(v);
        }
        return obj;
    }
    if node.as_i64().is_some() {
        return json::from(node.as_i64().unwrap());
    }
    if node.as_bool().is_some() {
        return json::from(node.as_bool().unwrap());
    }
    if node.as_str().is_some() {
        return json::from(node.as_str().unwrap());
    }
    return json::from("JSON ERROR"); // shouldn’t ever see this
}

pub fn parse_color(data: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut token = json::object! {
        id: id.to_owned(),
        "type": data["type"].as_str().unwrap(),
    };
    if is_alias(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else if !is_empty(&data["value"]) {
        let c = data["value"].as_str().unwrap();
        if HEX_RE.is_match(&c) {
            token["value"] = json_serialize(&data["value"]);
        } else {
            errors.push(format!("{}: invalid hex \"{}\"", id, &c));
        }
    } else {
        errors.push(format!("{}: missing value", id));
    }
    if !is_empty(&data["mode"]) {
        token["mode"] = json::JsonValue::new_object();
        for (k, v) in data["mode"].as_hash().unwrap() {
            let c = v.as_str().unwrap().trim().to_owned();
            if !HEX_RE.is_match(&c) {
                errors.push(format!("{}: invalid hex \"{}\"", id, &c));
            }
            token["mode"][k.as_str().unwrap().to_owned()] = json_serialize(v);
        }
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(token)
    };
}

pub fn parse_dimension(data: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut token = json::object! {
        id: id.to_owned(),
        "type": data["type"].as_str().unwrap(),
    };
    if is_alias(&data["value"]) {
        token["value"] = json_serialize(&data["value"])
    } else if !is_empty(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else {
        errors.push(format!("{}: missing value", id));
    }
    if !is_empty(&data["mode"]) {
        token["mode"] = json::JsonValue::new_object();
        for (k, v) in data["mode"].as_hash().unwrap() {
            token["mode"][k.as_str().unwrap()] = json_serialize(v);
        }
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(token)
    };
}

pub fn parse_font(data: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut token = json::object! {
        id: id.to_owned(),
        "type": data["type"].as_str().unwrap(),
        value: json::JsonValue::new_array(),
    };
    if is_alias(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else if !is_empty(&data["value"]) {
        // make sure token["value"] is always an array (data doesn’t have to be)
        if data["value"].is_array() {
            token["value"] = json_serialize(&data["value"]);
        } else {
            token["value"].push(json_serialize(&data["value"]));
        }
    } else {
        errors.push(format!("{}: missing value", id));
    }
    if !is_empty(&data["mode"]) {
        token["mode"] = json::JsonValue::new_object();
        for (k, v) in data["mode"].as_hash().unwrap() {
            if v.is_array() {
                token["mode"][k.as_str().unwrap()] = json_serialize(v);
            } else {
                token["mode"][k.as_str().unwrap()].push(json_serialize(v));
            }
        }
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(token)
    };
}

fn parse_cubic_bezier(data: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut token = json::object! {
        id: id.to_owned(),
        "type": data["type"].as_str().unwrap(),
        value: json::JsonValue::new_array(),
    };
    if is_alias(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else if !is_empty(&data["value"]) {
        match parse_cubic_bezier_item(&data["value"], id) {
            Err(msgs) => {
                for e in msgs {
                    errors.push(e);
                }
            }
            Ok(v) => token["value"] = v,
        }
    } else {
        errors.push(format!("{}: missing value", id));
    }
    if !is_empty(&data["mode"]) {
        token["mode"] = json::JsonValue::new_object();
        for (k, v) in data["mode"].as_hash().unwrap() {
            match parse_cubic_bezier_item(v, id) {
                Err(msgs) => {
                    for e in msgs {
                        errors.push(e);
                    }
                }
                Ok(b) => token["mode"][k.as_str().unwrap()] = b,
            }
        }
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(token)
    };
}

fn parse_cubic_bezier_item(value: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut cubic_bezier = json::JsonValue::new_array();
    if value.is_array() && value.as_vec().unwrap().len() == 4 {
        for (i, v) in value.as_vec().unwrap().iter().enumerate() {
            let mut num = 0.0;
            if v.as_f64().is_some() {
                num = v.as_f64().unwrap();
            } else if v.as_i64().is_some() {
                num = v.as_i64().unwrap() as f64;
            } else {
                let pos = ["x1", "y1", "x2", "y2"][i];
                errors.push(format!("{}: expected number for {}", id, pos));
                continue;
            }
            let is_x = i == 0 || i == 2;
            if is_x && (num < 0.0 || num > 1.0) {
                errors.push(format!(
                    "{}: x values must be between 0 and 1, received \"{}\"",
                    id, &num,
                ));
            }
            cubic_bezier.push(json::from(num));
        }
    } else {
        errors.push(format!("{}: expected format [x1, y1, x2, y2]", id));
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(cubic_bezier)
    };
}

pub fn parse_str(data: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut token = json::object! {
        id: id.to_owned(),
        "type": data["type"].as_str().unwrap()
    };
    if is_alias(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else if !is_empty(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else {
        errors.push(format!("{}: missing value", id));
    }
    if !is_empty(&data["mode"]) {
        token["mode"] = json::JsonValue::new_object();
        for (k, v) in data["mode"].as_hash().unwrap() {
            token["mode"][k.as_str().unwrap()] = json_serialize(v);
        }
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(token)
    };
}

pub fn parse_url(data: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut token = json::object! {
        id: id.to_owned(),
        "type": data["type"].as_str().unwrap()
    };
    if is_alias(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else if !is_empty(&data["value"]) {
        let v = data["value"].as_str().unwrap();
        if URL_RE.is_match(v) {
            token["value"] = json_serialize(&data["value"]);
        } else {
            errors.push(format!(
                "{}: invalid URL: \"{}\" (for relative files, use \"type: file\" instead)",
                id, v,
            ));
        }
    }
    if !is_empty(&data["mode"]) {
        token["mode"] = json::JsonValue::new_object();
        for (k, v) in data["mode"].as_hash().unwrap() {
            let url = v.as_str().unwrap();
            if URL_RE.is_match(url) {
                token["mode"][k.as_str().unwrap()] = json_serialize(v);
            } else {
                errors.push(format!(
                    "{}: invalid URL: \"{}\" (for relative files, use \"type: file\" instead)",
                    id, url,
                ));
            }
        }
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(token)
    };
}

pub fn parse_shadow(data: &Yaml, id: &String) -> Result<json::JsonValue, Vec<String>> {
    let mut errors: Vec<String> = Vec::new();
    let mut token = json::object! {
        id: id.to_owned(),
        "type": data["type"].as_str().unwrap()
    };
    if is_alias(&data["value"]) {
        token["value"] = json_serialize(&data["value"]);
    } else if !is_empty(&data["value"]) {
        if data["value"].is_array() {
            token["value"] = json_serialize(&data["value"]);
        } else {
            errors.push(format!(
                "{}: value must be array, received \"{}\"",
                id,
                json_serialize(&data["value"])
            ));
        }
    } else {
        errors.push(format!("{}: missing value", id));
    }
    if !is_empty(&data["mode"]) {
        token["mode"] = json::JsonValue::new_object();
        for (k, v) in data["mode"].as_hash().unwrap() {
            if v.is_array() {
                token["mode"][k.as_str().unwrap()] = json_serialize(v)
            } else {
                errors.push(format!(
                    "{}: value must be array, received \"{}\"",
                    id,
                    json_serialize(v)
                ));
            }
        }
    }
    return if errors.len() > 0 {
        Err(errors)
    } else {
        Ok(token)
    };
}

fn is_alias(data: &Yaml) -> bool {
    if !is_empty(data) && data.as_str().is_some() && ALIAS_RE.is_match(data.as_str().unwrap()) {
        return true;
    }
    return false;
}

fn is_empty(data: &Yaml) -> bool {
    return data.is_null() || data.is_badvalue();
}
